File: bindings/python/examples/custom_icon.py
#!/usr/bin/env python3
"""1:1 Python port of examples/custom-icon.c.

Creates a tray icon with a programmatically generated pixmap icon.

Run:
    GI_TYPELIB_PATH=build/src LD_LIBRARY_PATH=build/src \
        python3 bindings/python/examples/custom_icon.py
"""

import gi
gi.require_version("GdkPixbuf", "2.0")
gi.require_version("GLib", "2.0")
from gi.repository import GdkPixbuf, GLib

from gsni import StatusNotifierItem


def generate_tray_icon():
    size = 48
    half = size // 2
    stride = size * 4
    data = bytearray(stride * size)

    for y in range(size):
        for x in range(size):
            inside = (x - half) ** 2 + (y - half) ** 2 < half * half
            off = y * stride + x * 4
            if inside:
                data[off] = (0x50 + x * 180 // size) & 0xff
                data[off + 1] = (0x80 + y * 100 // size) & 0xff
                data[off + 2] = 0xcc
                data[off + 3] = 0xff
            else:
                data[off] = data[off + 1] = data[off + 2] = data[off + 3] = 0

    gb = GLib.Bytes.new(data)
    return GdkPixbuf.Pixbuf.new_from_bytes(
        gb, GdkPixbuf.Colorspace.RGB, True, 8, size, size, stride)


def main():
    icon = generate_tray_icon()
    item = StatusNotifierItem("custom-icon", title="Custom Icon Demo")
    item.status = 1  # ACTIVE
    item.icon_pixbuf = icon
    item.set_tooltip(None, "Custom Icon",
                     "Icon generated from raw pixel data")
    item.register()

    print("Tray icon registered with custom pixmap.")
    print("The icon is a programmatically generated colored circle.")
    print("Press Ctrl+C to exit.")

    loop = GLib.MainLoop.new(None, False)
    loop.run()


if __name__ == "__main__":
    main()

File: examples/custom-icon.c
/* SPDX-License-Identifier: LGPL-2.1-or-later
 * SPDX-FileCopyrightText: 2024 libgsni contributors
 *
 * Custom pixmap icon example — programmatically generates an icon
 * from raw pixel data and sets it via gsni_item_set_icon_pixbuf().
 */

#include <gio/gio.h>
#include "gsni/gsni.h"

static GdkPixbuf *
generate_tray_icon(void)
{
    gint size = 48;
    GdkPixbuf *pixbuf = gdk_pixbuf_new(GDK_COLORSPACE_RGB, TRUE, 8, size, size);
    guint8 *pixels = gdk_pixbuf_get_pixels(pixbuf);
    gint stride = gdk_pixbuf_get_rowstride(pixbuf);
    gint half = size / 2;

    for (gint y = 0; y < size; y++) {
        for (gint x = 0; x < size; x++) {
            gboolean inside_circle =
                (x - half) * (x - half) + (y - half) * (y - half)
                < half * half;

            guint8 *p = pixels + (gsize)y * stride + (gsize)x * 4;

            if (inside_circle) {
                p[0] = 0x50 + (guint8)(x * 180 / size); /* red   */
                p[1] = 0x80 + (guint8)(y * 100 / size); /* green */
                p[2] = 0xcc;                              /* blue  */
                p[3] = 0xff;                              /* alpha */
            } else {
                p[0] = p[1] = p[2] = 0x00;
                p[3] = 0x00; /* transparent */
            }
        }
    }

    return pixbuf;
}

int
main(int argc, char *argv[])
{
    g_autoptr(GApplication) app = g_application_new(
        "org.libgsni.CustomIcon", G_APPLICATION_DEFAULT_FLAGS);
    g_application_register(app, NULL, NULL);

    GDBusConnection *bus = g_application_get_dbus_connection(app);
    g_autoptr(GsniItem) item = gsni_item_new("custom-icon", bus, NULL);

    gsni_item_set_title(item, "Custom Icon Demo");
    gsni_item_set_status(item, GSNI_STATUS_ACTIVE);
    gsni_item_set_tooltip(item, NULL,
        "Custom Icon", "Icon generated from raw pixel data");

    /* Set icon as pixbuf instead of themed name */
    g_autoptr(GdkPixbuf) icon = generate_tray_icon();
    gsni_item_set_icon_pixbuf(item, icon);

    gsni_item_register(item, NULL);

    g_print("Tray icon registered with custom pixmap.\n"
            "The icon is a programmatically generated colored circle.\n"
            "Press Ctrl+C to exit.\n");

    g_autoptr(GMainLoop) loop = g_main_loop_new(NULL, FALSE);
    g_main_loop_run(loop);

    return 0;
}
